alf/algorithms/algorithm.py

```diff
@@ -1157,27 +1160,45 @@ def train_from_replay_buffer(self, update_global_counter=False):
 
         with record_time("time/train"):
             return self._train_experience(
-                experience, batch_info, num_updates, mini_batch_size,
+                experience,
+                batch_info,
+                num_updates,
+                mini_batch_size,
                 config.mini_batch_length,
                 (config.update_counter_every_mini_batch
-                 and update_global_counter))
-
-    def _train_experience(self, experience, batch_info, num_updates,
-                          mini_batch_size, mini_batch_length,
-                          update_counter_every_mini_batch):
+                 and update_global_counter),
+                whole_replay_buffer_training=config.
+                whole_replay_buffer_training)
+
+    def _train_experience(self,
+                          experience,
+                          batch_info,
+                          num_updates,
+                          mini_batch_size,
+                          mini_batch_length,
+                          update_counter_every_mini_batch,
+                          whole_replay_buffer_training: bool = False):
         """Train using experience."""
+        # Apply transformation and enrichment to the experience.
         experience = dist_utils.params_to_distributions(
             experience, self.experience_spec)
-        if self._exp_replayer_type != "one_time":
-            # The experience put in one_time replayer is already transformed
-            # in unroll().
-            experience = alf.data_structures.add_batch_info(
-                experience, batch_info, self._exp_replayer.replay_buffer)
-            experience = self.transform_experience(experience)
-            # allow data_transformers to change batch_info
+        experience = alf.data_structures.add_batch_info(
+            experience, batch_info, self._exp_replayer.replay_buffer)
+        experience = self.transform_experience(experience)
+
+        if whole_replay_buffer_training:
+            # For whole replay buffer training, explicitly set
+            # batch_info to None as an indicator for the following
+            # code to skip certain logics specifically designed for
+            # sampled replay buffer training.
+            batch_info = None
```

Comment:
I think we don't need to do this for whole_replay_buffer_training. line 1204 may still use it. Seems that batch_info is needed for line 1220. However, line 1220 seems unnecessary.

Reply:
I think we can always provide batch_info so that user don't need to worry about whether it is available or not.

Reply:
Sure. I added this because originally for `whole_replay_buffer_training` it is set to `None`, and when not set to `None`, training will crash (I did not investigate into it). Let me dig into it.